sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveHintsSuite.scala

```diff
@@ -295,4 +296,28 @@ class ResolveHintsSuite extends AnalysisTest {
         caseSensitive = true)
     }
   }
+
+  test("SPARK-35786: Support optimize repartition by expression in AQE") {
+    checkAnalysisWithoutViewWrapper(
+      UnresolvedHint("REBALANCE_PARTITIONS", Seq(UnresolvedAttribute("a")), table("TaBlE")),
+      RebalancePartitions(Seq(AttributeReference("a", IntegerType)()), testRelation))
+
+    checkAnalysisWithoutViewWrapper(
+      UnresolvedHint("REBALANCE_PARTITIONS", Seq.empty, table("TaBlE")),
+      RebalancePartitions(Seq.empty, testRelation))
+
+    withSQLConf(SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "false") {
+      checkAnalysisWithoutViewWrapper(
+        UnresolvedHint("REBALANCE_PARTITIONS", Seq(UnresolvedAttribute("a")), table("TaBlE")),
+        testRelation)
+
+      checkAnalysisWithoutViewWrapper(
+        UnresolvedHint("REBALANCE_PARTITIONS", Seq.empty, table("TaBlE")),
+        testRelation)
+    }
+
+    assertAnalysisError(
+      UnresolvedHint("REBALANCE_PARTITIONS", Seq(Literal(1)), table("TaBlE")),
```

Comment:
not related to this PR, I'm wondering if we should support `date(ts)` as the partition expression, instead of plain columns. cc @sunchao

Reply:
@cloud-fan yea I think it would be useful in general to support transforms as partition expressions. To achieve that, we'll need to first link the transforms with their actual semantics through function catalog (this is on my radar), and also allow plugin custom shuffle function in `ShuffleExchangeLike`. After that perhaps `RebalancePartitions` here can return `DataSourcePartitioning` when the partition expressions include transforms.